src/linker/Linker.Dataflow/TrimAnalysisPatternStore.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using Mono.Linker.Steps;
+
+namespace Mono.Linker.Dataflow
+{
+	public readonly struct TrimAnalysisPatternStore
+	{
+		readonly Dictionary<(MessageOrigin, bool), TrimAnalysisAssignmentPattern> AssignmentPatterns;
+		readonly Dictionary<MessageOrigin, TrimAnalysisMethodCallPattern> MethodCallPatterns;
+		readonly LinkContext _context;
+
+		public TrimAnalysisPatternStore (LinkContext context)
+		{
+			AssignmentPatterns = new Dictionary<(MessageOrigin, bool), TrimAnalysisAssignmentPattern> ();
+			MethodCallPatterns = new Dictionary<MessageOrigin, TrimAnalysisMethodCallPattern> ();
+			_context = context;
+		}
+
+		public void Add (TrimAnalysisAssignmentPattern pattern, bool isReturnValue)
+		{
+			// In the linker, each pattern should have a unique origin (which has ILOffset)
+			// but we don't track the correct ILOffset for return instructions.
+			// https://github.com/dotnet/linker/issues/2778
+			// For now, work around it with a separate bit.
+			AssignmentPatterns.Add ((pattern.Origin, isReturnValue), pattern);
+		}
```

Comment:
We should be able to deduce the "returnvalue" bit from the assignment, as the target should be MethodReturnValue.
Also in the linker the target can only be a single value 

No need for the additional parameter.